documentation/guides/migrating-from-v3-to-v4.md

```diff
@@ -0,0 +1,390 @@
+
+## Removed Exports/Components
+
+### WithContext
+
+`WithContext` as an internal abstraction used for class components. Use hooks or contextType instead.
+
+```jsx
+// old
+function Test({context}) {
+  console.log(context);
+  return null;
+}
+
+const TestWithContext = withContext(Consumer)(Test);
+
+// new with hooks
+function Test() {
+  const testContext = useContext(TestContext);
+  return <h1>{testContext}</h1>;
+}
+
+// new with contextType
+const TestContext = createContext('Polaris');
+
+class Test extends Component {
+  static contextType = TestContext;
+
+  render() {
+    const testContext = this.context;
+    return <h1>{testContext}</h1>;
+  }
+}
+```
+
+### WithRef
+
+`WithRef` was an internal component used to place refs on components wrapped in higher-order components. Use functional components instead.
+
+```jsx
+// old
+class Input extends React.PureComponent {
+  inputRef = createRef();
+
+  focus = () => {
+    inputRef.current.focus();
+  };
+
+  render() {
+    return <input ref={inputRef} {...props} />;
+  }
+}
+
+const WrappedInput = compose(
+  hocOne,
+  hocTwo,
+  hocX,
+  withRef,
+)(Input);
+
+class App extends React.PureComponent {
+  input = createRef();
+
+  componentDidMount() {
+    input.current.focus();
+  }
+
+  render() {
+    return (
+      <ReffableInput
+        ref={reffableInput}
+        onFocus={() => console.log('focused')}
+      />
+    );
+  }
+}
+
+// new with hooks
+function Input(props, ref) {
+  const inputRef = useRef();
+  useImperativeHandle(ref, () => ({
+    focus: () => {
+      inputRef.current.focus();
+    },
+  }));
+  return <input ref={inputRef} {...props} />;
+}
+const ReffableInput = forwardRef(Input);
+
+function App() {
+  const reffableInput = useRef();
+
+  useEffect(() => {
+    reffableInput.current.focus();
+  }, []);
+
+  return (
+    <ReffableInput ref={reffableInput} onFocus={() => console.log('focused')} />
+  );
+}
+```
+
+### Autocomplete.ComboBox.TextField
+
+Use `Autocomplete.TextField` instead.
+
+```jsx
+...
+const textFieldMarkup = (
+  <Autocomplete.TextField
+    label="Example"
+  />
+);
+...
+```
+
+### Autocomplete.ComboBox.OptionList
+
+Use `OptionList` instead.
+
+```jsx
+...
+const optionListMarkup = (
+  <OptionList
+    onChange={() => {}}
+    selected={[]}
+  />
+);
+...
+```
+
+### Navigation.UserMenu
+
+Use `TopBar.UserMenu` instead
+
+```jsx
+const userMenuMarkup = (
+  <TopBar.UserMenu
+    actions={[]}
+    name="Andrew"
+    initials="AM"
+    open
+    onToggle={() => {}}
+  />
+);
+```
+
+### Modal.Dialog/Tabs.Panel
+
+`Dialog` and `Panel` are subcomponents that are used by Polaris to build the respective component and are not intended to be used outside of their component.
+
+## Dependencies
+
+React/ReactDOM are peer dependencies of Polaris and have been bumped to 16.8.6 to enable the use of hooks. Use `yarn` or `npm` to install a recent version of react.
+
+```bash
+# yarn
+yarn add react react-dom
+
+# npm
+npm i react react-dom
```

Comment:
Doesn't this line require `--save-dev` or `--save` by any chance?

Reply:
Packages installed by npm will be added as dependencies by default `npm i --save X === npm i X`